### PR TITLE
Credentials: partial migration to SDK v3

### PR DESCRIPTION
**Pull request by JadenSimon**
## Problem
SDK v2 has some known issues in regards to credentials from `source_profile`. SDK v3 does not fix them out-right, however, it does expose some functionality that allows us to fix the bugs.

## Solution
Migrate the credentials subsystem to use SDK v3, then add additional logic to handle `source_profile` resolution.
The new SDK credentials API allows us to 'pre-load' credentials when calling relevant APIs. We can use this functionality to load source profile credentials ourselves, letting the SDK treat them as static credentials. Any valid profile should now be usable as a base profile as well (e.g. processes, SSO, MFA, etc.)

~TODO: re-add test for when SDK never resolves a credential provider. Current test broke since the API is a little bit different with v3~
~TODO: check if the STS region hack is still needed~
~TODO: add regression tests~

### Extra testing
Still thinking of more tests I can add. 

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by JadenSimon** [comment]
Ignore the huge `package-lock` diff for now, it's caused by a force-update due to peer dependencies unrelated to this PR.

**Comment by JadenSimon** [comment]
SDK has broken types I guess.

**Comment by JadenSimon** [comment]
Still need to test for C9 and some edge cases. Will be adding a few more unit tests too.

File: src/credentials/providers/sharedCredentialsProvider.ts [hunk 1]
```diff
@@ -141,27 +150,50 @@ export class SharedCredentialsProvider implements CredentialsProvider {
...
-
-        try {
-            // Profiles with references involving non-aws partitions need help getting the right STS endpoint
-            this.applyProfileRegionToGlobalStsConfig()
-            //  SSO entry point
-            if (this.isSsoProfile()) {
-                const ssoCredentialProvider = this.makeSsoProvider()
-                return await ssoCredentialProvider.refreshCredentials()
+    // Patches 'source_profile' credentials as static representations, which the SDK can handle in all cases
+    private async patchSourceCredentials(): Promise<ParsedIniData> {
```
Comment (JadenSimon) [inline src/credentials/providers/sharedCredentialsProvider.ts:154]: This part of the PR fixes issues with the JS SDK's credential handling. The SDK is unable to resolve `source_profile` fields when the source profile uses SSO/MFA/`credential_process`. We can handle this resolution ourselves, giving the SDK the resolved credentials by 'pre-loading' them.

File: src/credentials/providers/sharedCredentialsProvider.ts [hunk 2]
```diff
@@ -227,82 +259,89 @@ export class SharedCredentialsProvider implements CredentialsProvider {
...
 
-    private makeSharedIniFileCredentialsProvider(): () => AWS.Credentials {
-        return () =>
-            new AWS.SharedIniFileCredentials({
-                profile: this.profileName,
-                tokenCodeFn: async (mfaSerial, callback) =>
-                    await getMfaTokenFromUser(mfaSerial, this.profileName, callback),
-            })
+    private makeSharedIniFileCredentialsProvider(loadedCreds: ParsedIniData): AWS.CredentialProvider {
+        const assumeRole = async (credentials: AWS.Credentials, params: AssumeRoleParams) => {
```
Comment (JadenSimon) [inline src/credentials/providers/sharedCredentialsProvider.ts:294]: V3 no longer performs role assumption for us; we must do it manually.

File: src/test/credentials/provider/ec2CredentialsProvider.test.ts [hunk 3]
```diff
@@ -78,11 +77,6 @@ describe('Ec2CredentialsProvider', function () {
         assert.strictEqual(credentialsProvider.getDefaultRegion(), undefined)
     })
 
-    it('returns credentials', async function () {
-        const credentials = await credentialsProvider.getCredentials()
-        assert(credentials instanceof EC2MetadataCredentials)
```
Comment (JadenSimon) [inline src/test/credentials/provider/ec2CredentialsProvider.test.ts:83]: V3 credentials are no longer instances of a class, they are just plain old data instead. So this test is unnecessary.

File: src/credentials/providers/sharedCredentialsProvider.ts (hunk 1 above)
Comment (breedloj) [reply src/credentials/providers/sharedCredentialsProvider.ts:154]: Has this come up as a case customers are using? Do we have any idea if the ASK CLI supports this? Seems like a nice improvement on paper but we should also tread carefully when adding custom credentials logic on top of what the SDK provides. It will increase the surface area for bugs to creep in (and credentials always seem to have goofy edge cases) and is maintainability overhead. Another approach is contributing changes back to the JS SDK if we feel this is a gap.

Comment (breedloj) [reply src/credentials/providers/sharedCredentialsProvider.ts:154]: Makes sense, thanks for the context. Agree this is the right solution for now since there's active customer pain on this issue.

**Review by breedloj — APPROVED** [review]

File: plugins/isolation/index.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
+"use strict";
```
Comment (justinmk3) [inline plugins/isolation/index.js]: this this JS file intentionally committed?

File: src/credentials/providers/sharedCredentialsProvider.ts [hunk 5]
```diff
@@ -141,27 +150,50 @@ export class SharedCredentialsProvider implements CredentialsProvider {
...
-        const originalStsConfiguration = AWS.config.sts
-
-        try {
-            // Profiles with references involving non-aws partitions need help getting the right STS endpoint
-            this.applyProfileRegionToGlobalStsConfig()
-            //  SSO entry point
-            if (this.isSsoProfile()) {
-                const ssoCredentialProvider = this.makeSsoProvider()
-                return await ssoCredentialProvider.refreshCredentials()
+    // Patches 'source_profile' credentials as static representations, which the SDK can handle in all cases
```
Comment (justinmk3) [inline src/credentials/providers/sharedCredentialsProvider.ts:153]: i think an extended comment is justified here. Just pasting your comment from this PR would be good:
```suggestion
    // Patches 'source_profile' credentials as static representations, which the SDK can handle in all cases.
    //
    // The SDK is unable to resolve source_profile fields when the source profile uses SSO/MFA/credential_process.
    // We can handle this resolution ourselves, giving the SDK the resolved credentials by 'pre-loading' them.
```

File: plugins/isolation/index.js (hunk 4 above)
Comment (JadenSimon) [reply plugins/isolation/index.js]: Was some experimental code, surprised it was showing that commit for you since I made it like 3 months ago.

File: src/test/testUtil.ts [hunk 6]
```diff
@@ -83,3 +83,14 @@ export async function tickPromise<T>(promise: Promise<T>, clock: FakeTimers.Inst
     clock.tick(t)
     return await promise
 }
+
+/** Recursively delete undefined key/value pairs */
+export function stripUndefined(obj: any): void {
```
Comment (justinmk3) [inline src/test/testUtil.ts:88]: this could live in collectionUtils

**Review by justinmk3 — APPROVED** [review]

File: src/test/testUtil.ts (hunk 6 above)
Comment (JadenSimon) [reply src/test/testUtil.ts:88]: Yeah probably better there

**Review by justinmk3 — CHANGES_REQUESTED** [review]
checking ec2 instance credentials

**Review by justinmk3 — APPROVED** [review]
